### PR TITLE
fix(renderer): give pickRect the batched-mesh path pick() already has (#1904)

### LINKED ISSUE #1904: Ctrl + click +drag rectangle-select returns no selection, `pickRect` returns empty on large models
### Version

ifc-lite v1.33.0

### Which surface

Viewer (browser)

### Reproduction

Bug found by me, details and a patch proposed by Claude

## Steps to reproduce 

1. Load a large model (large enough to trip the >500 pick-mesh threshold — see cause 2).
2. Hold Ctrl and drag a box over several elements.
3. Release.

**Expected:** every element under the box is selected.
**Actual:** nothing is selected (or exactly one element — whatever was under the cursor
at release).

`pickRect` returns empty on large models

`PickingManager.pick()` switches to CPU raycasting once a model needs more than
`MAX_PICK_MESH_CREATION = 500` pick meshes (`picking-manager.ts:154-165`), precisely
because "GPU picking requires individual mesh buffers; for 60K+ elements this is too
slow." As a result the per-element GPU pick meshes are **never hydrated**.

`pickRect` has no equivalent fallback. This is already documented at
`picking-manager.ts:236-240`:

> *Limitations: skips the CPU-raycast and dynamic-mesh-creation fallbacks that `pick()`
> uses for very large batched models, so rect pick may miss entities whose individual
> meshes haven't been hydrated. Acceptable for an MVP — rect select is a power-user tool
> and the user can fall back to single-click pick.*

In practice this isn't "may miss some entities" — on any model past the 500-mesh
threshold the GPU pick pass renders nothing at all and rect select misses **everything**.
That's also why Ctrl+click still works (CPU path) while Ctrl+drag doesn't (GPU path).

**Measured on the same drag:** `pickRect` → `count: 0`; a CPU AABB fallback → **108
entities**.

### Suggested fix

Give `pickRect` a CPU fallback mirroring the one `pick()` already has. What worked for me:
when the GPU pass returns empty, project each entity's cached world AABB through the
camera `viewProj` matrix and keep the entities whose projected screen rect intersects the
drag rect. `Scene.getAllMeshDataExpressIds()` + `Scene.getEntityBoundingBox()` are
sufficient — both work without GPU hydration and even after geometry release.

Sketch:

```js
const ids = scene.getAllMeshDataExpressIds();
const m = camera.getViewProjMatrix().m; // column-major: element(row,col) = m[col*4+row]

for (const id of ids) {
  if (hiddenIds?.has(id)) continue;
  if (isolatedIds && !isolatedIds.has(id)) continue;

  const bb = scene.getEntityBoundingBox(id);
  if (!bb) continue;

  // project the 8 AABB corners, skipping any with w <= 0 (behind camera),
  // take the screen-space bounds, and test intersection against the drag rect
}
```

Two notes on semantics:

- An AABB **intersection** test is more permissive than exact per-texel coverage — a box
  clipping the corner of a large diagonal element will select it. That matches typical CAD
  marquee behaviour and is the same trade-off `pick()` already makes with AABB-filtered
  raycasting, but it's a deliberate choice worth confirming. A containment test would be
  the stricter alternative.
- Applying the `hiddenIds` / `isolatedIds` filters keeps the existing invariant that you
  can never select something you can't see.

Gating the fallback on "GPU pass returned empty" keeps exact GPU picking for models below
the threshold and costs nothing there.

---

## Related, minor: docstring says Shift+drag

`Picker.pickRect`'s docstring (`packages/renderer/src/picker.ts:476`) says *"Used by the
Shift+drag rectangle-selection UI"*, but the binding is Ctrl/⌘+drag
(`useMouseControls.ts:470`). Shift+drag is pan in the select tool. Harmless, but it sent
me looking in the wrong place initially.

### Expected vs actual

**Expected:** every element under the box is selected.
**Actual:** nothing is selected.

### Environment

Windows 11
Microsoft Edge Version 150.0.4078.99

### Privacy

- [x] My attached files and text contain no confidential or client-identifying data.

### PR DESCRIPTION
**Pull request by BIMvoice**
## What and why

Closes #1904. Thanks @louistrue for the diagnosis and browser repro — this implements it.

`PickingManager.pickRect` passed `Scene.getMeshes()` straight to the GPU pick pass. Batched geometry lives in `batchedMeshes` and never reaches that list, so rectangle select returned an empty set deterministically.

`pick()` has had three things `pickRect` lacked:
1. a CPU fallback once JS geometry data is released,
2. a CPU fallback when hydrating one mesh per visible piece would exceed `MAX_PICK_MESH_CREATION`,
3. the hydration itself.

That decision now lives in **one** place — `prepareBatchedPick()` — used by both paths. Duplicating those ~40 lines would have recreated exactly the divergence class `AGENTS.md` warns about ("a fix applied to only one pipeline silently diverges").

**One scoping note worth your call.** The issue suggests "the batched-mesh path (and/or the CPU-raycast fallback)". Those aren't interchangeable here: hydration alone closes the `hello-wall.ifc` repro but leaves the large-model case in the title broken, because past the budget the code *must* go to CPU and there was no rectangle equivalent of `Scene.raycast`. So this adds `Scene.selectRect()`, mirroring `Scene.raycast()`, backed by a new `scene-rect-select` module that projects entity AABBs through the view-projection and tests screen-rect overlap.

**Known trade-off, flagged rather than buried:** the CPU rect path is bounding-box granular, so it can over-select slightly versus the pixel-exact GPU pass — an entity is selected when the rect overlaps its bounds even if it covers empty space inside them. That is the fidelity the released-geometry raycast path already has, and over-selection is recoverable by the user where selecting nothing is not. If you would rather it did a triangle-level test on the non-released path, say so and I will add it. Hidden / isolated / section-plane / crop-box filtering all apply, so selection still matches what is visible; instanced-only occurrences are covered since their AABBs are already registered in `boundingBoxes`.

## How it was verified

- **Proved the regression tests catch the bug**: reverting `pickRect` to the pre-fix code path fails all four new cases (`hydrates batched pieces…`, `falls back to Scene.selectRect once geometry data is released`, `…when hydration would exceed the pick-mesh budget`, `honours isolation…`); restoring the fix returns them to green.
- `pnpm --filter @ifc-lite/renderer test` — **399 passed, 0 failed** (384 before this change, so the `pick()` extraction is behaviour-preserving, +15 new).
- `pnpm test` — 74/74 turbo tasks.
- `pnpm typecheck` — 33/33.
- `node scripts/check-test-wiring.mjs` — clean.
- `node scripts/check-api-surface.mjs` — clean, no snapshot update needed (the snapshot tracks exported names; `Scene.selectRect` is a new member on an already-exported class).

New coverage: 4 cases in `picking-manager.test.ts` for the batched routing, and 11 in `scene-rect-select.test.ts` pinning the projection itself — rect overlap, partial overlap, drag-direction normalisation, hidden/isolated filtering (including empty-isolation meaning "nothing selectable"), geometry behind the camera, a box straddling the camera plane, degenerate viewport, and crop-box exclusion.

No Rust touched. `pnpm build:wasm` + `pnpm test:wasm-contract` (36/36) do pass on this branch and the committed `pkg/ifc-lite.d.ts` shows no drift.

## Checklist

- [x] A test asserts the new behavior through a fixture or a stated invariant — verified to fail before the fix
- [x] Published `packages/*` touched: changeset added (`@ifc-lite/renderer` minor — `Scene.selectRect` is additive public surface). Export surface unchanged, so no `pnpm api-surface:update`
- [x] No geometry output changed — determinism manifests untouched
- [x] House rules: no `as any` / `@ts-ignore`, no silent `catch {}`, `picking-manager.ts` is 300 lines, no repo-wide `cargo fmt`
- [x] No client or project identifiers in code, tests, commit messages, or this PR

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved rectangle selection for batched models, including released or incomplete geometry.
  * Added reliable CPU fallback selection with camera clipping and viewport handling.
  * Preserved visibility, isolation, and crop-box filters.
  * Maintained point-cloud selection during fallback, with graceful handling when point data is unavailable.
  * Improved selection of partially visible and camera-straddling objects.

* **Tests**
  * Added coverage for GPU hydration, CPU fallback, projection, partial overlap, drag direction, hidden entities, clipping, and viewport edge cases.

### CODE REVIEW

**Comment by louistrue** [comment]
Heads-up: I pushed a follow-up commit (e5bbe3f8) directly to this branch, so **please `git pull` before you continue working on it**. Sorry for the surprise. I saw you pushing to #1930 at the same time and did not want you to hit a conflict or redo work.

First, the review itself. I checked your claims rather than taking them, and every number reproduced exactly: 399 passing (384 on main), and reverting `pickRect` really does fail 4 of the 6 picking-manager cases. The `prepareBatchedPick()` extraction is a genuine verbatim lift, same ordering, same guards, same budget accounting, so the single-click `pick()` path is not at risk. That was the thing most likely to go wrong in this change and it did not. Good instinct putting the shared decision in one place.

What the follow-up commit changes:

**1. Point clouds regressed against main.** This is the one real defect. On main, `pickRect` always reached `picker.pickRect` with the point snapshot, and splats render into the pick pass whether or not per-element *mesh* buffers were hydrated. So point-cloud rect selection worked on main, and the new `'cpu'` branch returned `Scene.selectRect` before ever consulting `pointPickProvider`. The CPU branch now runs a point-only pass and unions the hits, degrading to the box hits with a log if readback fails.

Greptile flagged this too but framed it as an asymmetry with `pick()`. That framing is wrong: `pick()`'s CPU fallback (`scene.raycast`) never included point clouds either. The regression is specifically against main's `pickRect`.

**2. Near-plane clipping.** Your `straddlesCamera` branch added any box with a corner behind the camera plane to *any* rect anywhere on screen. On the large models this path exists to serve, the user is usually standing inside the building, so the floor, ceiling and surrounding walls all straddle. A 5px drag over a chair grabbed all of them. Boxes are now clipped against the near and far planes before projecting. Also added the missing ndc-z test.

**3. Prose that overclaimed.** The JSDoc had the selectivity direction backwards (it said "more selective"; bounding-box granularity and no depth test both make it *less* selective), two comments credited the raycast path with the same no-depth-test behaviour when `scene-raycaster.ts:341` does depth-order by bounding-box entry distance, and the changeset said two remaining divergences when there are three.

On CodeRabbit's perf complaint about the bounding-box warm loop: I looked and it is a false alarm as a *new* regression. `scene-raycaster.ts:377-388` already calls `getEntityBoundingBox` for every entity on the first CPU single-click pick, same cost class, same population, shipped long ago. I left the loop alone.

Renderer suite is 407 passing, 0 failures.

One pattern worth naming, because it showed up in all three of your PRs and it is the only thing separating good work from finished work here: you verify the path you build, but not the path you displace. Each blocker was a new early-return skipping something the old code reached. Worth asking of every new branch: what did the old code get to that this now skips?

**Comment by louistrue** [comment]
@coderabbitai review

File: packages/renderer/src/picking-manager.ts [hunk 1]
```diff
@@ -256,6 +266,17 @@ export class PickingManager {
...
+            return this.scene.selectRect(
+                sx0, sy0, sx1, sy1,
+                this.canvas.width, this.canvas.height,
+                this.camera.getViewProjMatrix().m,
+                options?.hiddenIds,
+                options?.isolatedIds,
+                clip,
+            );
+        }
+
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/renderer/src/picking-manager.ts:335]: Fixed in e5bbe3f8. The CPU branch now runs a point-only pass and unions its hits with `Scene.selectRect`, so splats are selected again on this path.

One correction to the framing though: this was described as an asymmetry with `pick()`. It is not. `pick()`'s CPU fallback is `scene.raycast`, which has never included point clouds either, so the two were already symmetric. The real regression was against **main's `pickRect`**, which always reached `picker.pickRect` with the point snapshot. Splats render into the pick pass whether or not per-element mesh buffers were hydrated, so point selection worked on main and the new early return dropped it.

File: packages/renderer/src/scene-rect-select.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            if (!(cw > 1e-6)) {
+                straddlesCamera = true;
+                continue;
+            }
+            const cx = m[0] * x + m[4] * y + m[8] * z + m[12];
+            const cy = m[1] * x + m[5] * y + m[9] * z + m[13];
+
+            // NDC -> canvas pixels. Y flips: NDC is up-positive, screen is down-positive.
+            const sx = ((cx / cw) * 0.5 + 0.5) * viewportWidth;
+            const sy = (0.5 - (cy / cw) * 0.5) * viewportHeight;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/renderer/src/scene-rect-select.ts]: Fixed in e5bbe3f8. `scene-rect-select.ts` now clips each box against the near and far planes before projecting and tests normalized depth, so an entity outside the depth slab no longer passes on its x/y bounds alone.

That also replaced the bigger problem next to it: the old code added any box with a corner behind the camera plane to every rect anywhere on screen, which on the large models this path serves (camera inside the building) selected the floor, ceiling and surrounding walls on a 5px drag.

File: packages/renderer/src/scene.ts [hunk 3]
```diff
@@ -3958,4 +3959,50 @@ export class Scene {
...
+  ): Set<number> {
+    // After release the cache is already the complete set; before it, boxes are
+    // computed lazily, so make sure every entity that still has mesh data has
+    // one. Same authoritative id set pick() uses, so colour-fused fillers whose
+    // id lives only in per-vertex entityIds are not skipped (#1358).
+    if (!this.geometryReleased) {
+      for (const expressId of this.getAllMeshDataExpressIds()) {
+        this.getEntityBoundingBox(expressId);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/renderer/src/scene.ts:4005]: Checked this and I do not think it is a new regression, so I have left the loop alone.

`scene-raycaster.ts:377-388` already calls `getEntityBoundingBox` for every entity on the first CPU single-click pick. Same cost class, same over-budget/large-model population, and it has shipped for a long time. The cache is shared and one-time, so the first Ctrl+drag pays it only if no prior pick primed it.

Measured rather than assumed: 60k entities x 400 verts is ~87ms one-time, and ~11ms per subsequent drag. Worth knowing, and the comment now says so honestly instead of claiming it costs nothing, but it is not something this PR introduced.

File: .changeset/brown-pianos-clap.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+- Bounding-box granularity, so a rect covering only empty space inside an element's bounds still selects it.
+- No depth test, so it selects through occlusion, where the GPU rect pass only ever sees front-most fragments.
+- Whole-box section-plane and crop-box filtering: the CPU path skips an entity only when nothing of its box could be visible, whereas the pick shader discards per fragment, so a rect over the sectioned-away half of a box still selects it.
+
+Hidden and isolation filtering do apply, per entity, exactly as on the GPU path.
+
+Point clouds keep working on this path: splats render into the pick pass whether or not per-element mesh buffers were hydrated, so the CPU branch still runs a point-only GPU pass and unions its hits into the bounding-box result. Those point hits carry the section plane (the point picker clips on it) but not the crop box, and not the hidden/isolated sets — unchanged from the existing GPU rect pass, which has never filtered point nodes by those sets. If that point pass fails at readback, the rectangle select degrades to the bounding-box hits and logs, rather than failing outright.
```
[the remark this answers is not in the document]
Comment (louistrue) [reply .changeset/brown-pianos-clap.md:19]: Good catch, and I took the second option you offered: keep the behaviour and make it explicit.

Point splats are deliberately not filtered by `hiddenIds`/`isolatedIds` on either branch. Per-asset point visibility is handled upstream and the pick pass has never filtered point nodes by the entity-level sets, so filtering only the new CPU union would have made the two rect paths disagree with each other.

2f7b4bb3 adds a test pinning exactly that, and I mutation-checked it: filtering splats by hidden/isolated fails that one case and nothing else (407/408). So changing this later becomes a deliberate decision against a red test rather than a silent behaviour change. The changeset already described the divergence; the test now enforces it.

**Comment by louistrue** [comment]
@coderabbitai review

File: packages/renderer/src/picking-manager.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (BIMvoice) [reply packages/renderer/src/picking-manager.ts:335]: Acknowledged, and the correction matters more than the fix.

The asymmetry-with-`pick()` framing was mine, not just Greptile's — I checked whether `pick()`'s CPU fallback handled point clouds, found it didn't, concluded "pre-existing, not a regression," and downgraded the finding. Then I explicitly briefed the work as documentation-only, no implementation. So the wrong framing didn't just travel with the report; it decided what got fixed.

The right comparison was `pickRect`'s own behaviour on main, not a sibling function's. That is now a standing rule for me: for a regression check, compare against the previous behaviour of the function being changed.

Verified your commit locally rather than assuming: 407 passing, and `scene-raycaster.ts:377` does already warm every bounding box on the first CPU click, so CodeRabbit's perf finding really was a false alarm as you said. I had an uncommitted batch-bounds pre-filter for it and dropped it — near/far clipping supersedes my straddle-plus-eye heuristic anyway, and it is the fix I described as "more involved but standard" and then didn't do. Nothing from me is pending on this branch.
